File: prototypes/docs-rag-qa/typescript/src/tools/chunker.ts
/**
 * Document chunking utility.
 */

export function chunkDocument(
  content: string,
  chunkSize = 500,
  overlap = 50,
): string[] {
  if (!content.trim()) return [];

  const sentences = content.match(/[^.!?]+[.!?]+\s*/g) ?? [content];
  const chunks: string[] = [];
  let current = "";

  for (const sentence of sentences) {
    if (current.length + sentence.length > chunkSize && current.length > 0) {
      chunks.push(current.trim());
      // Keep overlap from end of previous chunk
      const words = current.split(/\s+/);
      const overlapWords = words.slice(
        Math.max(0, words.length - Math.ceil(overlap / 5)),
      );
      current = `${overlapWords.join(" ")} ${sentence}`;
    } else {
      current += sentence;
    }
  }

  if (current.trim()) {
    chunks.push(current.trim());
  }

  return chunks;
}

File: prototypes/docs-rag-qa/python/app/tools/chunker.py
"""Document chunking utilities."""

import re


def chunk_document(content: str, chunk_size: int = 500, overlap: int = 50) -> list[str]:
    """Split text into overlapping chunks by sentence boundaries.

    Args:
        content: The text to split into chunks.
        chunk_size: Target size for each chunk in characters.
        overlap: Number of characters to overlap between chunks.

    Returns:
        A list of text chunks.
    """
    if not content or not content.strip():
        return []

    # Split into sentences using common delimiters
    sentences = re.split(r"(?<=[.!?])\s+", content.strip())
    sentences = [s for s in sentences if s.strip()]

    if not sentences:
        return []

    chunks: list[str] = []
    current_chunk = ""

    for sentence in sentences:
        candidate = f"{current_chunk} {sentence}".strip() if current_chunk else sentence

        if len(candidate) > chunk_size and current_chunk:
            chunks.append(current_chunk)
            # Apply overlap: take the tail of the current chunk
            if overlap > 0 and len(current_chunk) > overlap:
                current_chunk = current_chunk[-overlap:] + " " + sentence
            else:
                current_chunk = sentence
        else:
            current_chunk = candidate

    if current_chunk.strip():
        chunks.append(current_chunk.strip())

    return chunks
